src/keyboardNavigation.ts

```diff
@@ -0,0 +1,83 @@
+        return findNearestParentWithAdditionalLeaves(parent);
+    } else {
+        return <HTMLElement>parentAndSiblings[parentAndSiblings.indexOf(parent) + 1];
+    }
+}
+
+export const navigateUp = (path: string, shadowRoot: ShadowRoot): void => {
+    try {
+        const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+        if (!currentFocus || !currentFocus.parentNode) return;
+        const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+        var previousSibling: HTMLElement = <HTMLElement>(
+            currentFocus.parentNode.children[currentFocusAndSiblings.indexOf(currentFocus) - 1]
+        );
+        if (!previousSibling) {
+            //deal with the item being the first entry in a UL
+            previousSibling = <HTMLElement>currentFocus.parentNode.parentNode;
+        }
+        if (
+            !isDeepestVisible(previousSibling) &&
+            !isImmediateParent(path, previousSibling.getAttribute('data-path') || '')
+        ) {
+            changeFocusedElement(findDeepestVisibleChild(previousSibling), currentFocus);
+        } else {
+            changeFocusedElement(previousSibling, currentFocus);
+        }
+    } catch (err) {
+        console.error('Failed to navigate up via key press');
+    }
+};
+
+export const navigateDown = (path: string, shadowRoot: ShadowRoot): void => {
+    try {
+        const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+        if (!currentFocus || !currentFocus.parentNode) return;
+        const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+        if (!isDeepestVisible(currentFocus)) {
+            changeFocusedElement(<HTMLElement>currentFocus.children[1].children[0], currentFocus);
+        } else if (currentFocusAndSiblings.indexOf(currentFocus) === currentFocusAndSiblings.length - 1) {
+            changeFocusedElement(findNearestParentWithAdditionalLeaves(currentFocus), currentFocus);
+        } else {
+            var nextSibling: HTMLElement = <HTMLElement>(
+                currentFocus.parentElement?.children[currentFocusAndSiblings.indexOf(currentFocus) + 1]
+            );
+            changeFocusedElement(nextSibling, currentFocus);
+        }
+    } catch (err) {
+        console.error('Failed to navigate down via key press');
+    }
+};
+
+export const navigateToParent = (path: string, shadowRoot: ShadowRoot): void => {
+    try {
+        const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+        if (!currentFocus || !currentFocus.parentNode || !currentFocus.parentNode.parentNode) return;
+        changeFocusedElement(<HTMLElement>currentFocus.parentNode.parentNode, currentFocus);
+    } catch (err) {
+        console.error('Failed to navigate to parent via key press');
+    }
+};
+
+const changeFocusedElement = (newFocus: HTMLElement, oldFocus: HTMLElement): void => {
```

Comment:
i really want to suggest not crawling "down one level" but i know it's totally uniform (picking the `span`s inside of the `li`s, right?). but since you're not crawling down, i think it'd really help for this function to be clearly named; something like `changeChildFocus` or something that indicates that the passed elements aren't actually getting focus effects

Reply:
Yeah, crawling down one level is the best thing I landed on to standardize what sort of elements this was prepared to handle. Previously I was passing around spans and list items, and doing some really horrible logic to handle it here. Now I only deal with list items, and found this way to actually focus the correct element. 

I'm going with `changeFocusedKey` for now. I think Alex's comment
> Instead of children[x] could we do node.querySelector('[part="object"]') and/or node.querySelector('[part="preview"]') instead?

Might uncover a better way of grabbing elements based on their part rather than indexing, but I'm saving that for the end since theres a strong change I blow up the world. 

Changes applied

Reply:
Yeah, after [f9eb614](https://github.com/deepgram/json-viewer/pull/1/commits/f9eb614f98a6680b1f6c01600178362c244c550f) I'm fairly happy with this decision